grails-app/services/org/pih/warehouse/invoice/PrepaymentInvoiceMigrationService.groovy

```diff
@@ -0,0 +1,83 @@
+package org.pih.warehouse.invoice
+
+import grails.gorm.transactions.Transactional
+
+import org.pih.warehouse.core.Constants
+import org.pih.warehouse.order.OrderAdjustment
+import org.pih.warehouse.order.OrderItem
+import org.pih.warehouse.order.OrderItemStatusCode
+
+/**
+ * This service exists for the sole purpose of performing data migrations for changelog 0.9.x/2024-08-20-0000.
+ * The logic was only put in a service class so that it could be unit tested. Do not use it in regular flows.
+ */
+@Transactional
+class PrepaymentInvoiceMigrationService {
+
+    /**
+     * For STEP 1 of the migration.
+     *
+     * Set the amount field for all prepayment invoice items that existed before the partial invoicing
+     * feature was introduced.
+     */
+    void updateAmountForPrepaymentInvoiceItems() {
+        InvoiceType prepaymentInvoiceType = InvoiceType.findByCode(InvoiceTypeCode.PREPAYMENT_INVOICE)
+
+        Invoice.findAllByInvoiceType(prepaymentInvoiceType).each { Invoice prepaymentInvoice ->
+            // The amount field will be null for all pre-existing invoices since the field was not being used.
+            prepaymentInvoice.invoiceItems.find { it.amount == null }.each { InvoiceItem prepaymentInvoiceItem ->
+                prepaymentInvoiceItem.amount = computePrepaymentInvoiceItemAmount(prepaymentInvoiceItem)
+                prepaymentInvoiceItem.save(failOnError: true, flush: true)
```

Comment:
Is this flush needed?

Reply:
I was thinking there'd be some GORM weirdness since we're updating these invoice items and then fetching them pretty soon after in step 2 of the migration but I'll test it out without the flush. Maybe GORM can handle the case gracefully. If not, maybe there's a way to just flush the cache at the end.

It's also making me wonder if GORM is batching these saves together or if we'll need to do it explicitly.

I think all this is highlighting the fact that I know very little about how GORM works behind the scenes...